Initial Haddock support.

Fuuzetsu opened the pull request:
I'm able to generate documentation for existing
library (`haddock-lib-b` in tests) that depends on another. They both
automagically get their docs generated. Some notes:

* Linking to direct dependencies seems to work. I'm unsure that we
  need transitive interface file inclusion: I don't think so.
* It's able to skip non-Haskell dependencies so it doesn't try to do
  something weird on system deps for example but see comment in tests.
* Linking to `prebuilt_dependencies` documentation is not solved: we'd
  need to find out the actual paths to these to pass to Haddock. A
  possibility is to get it from GHC itself, basically a little script
  that calls `ghc-pkg`, finds out where packages live and gets the
  paths that way. Possibly we don't care about providing this with
  "you should be defining dependencies in bazel". I think we should
  provide this if it's not too hard though.
* We only output default files for the direct dependencies mentioned
  in `haskell_haddock`. This seems to work OK though I'm not 100% on
  whether we can rely no other docs always being neighbours. See
  comment in code on `--read-interface`. FWIW we could output them all
  if we wanted but it doesn't change the assumption.
* I haven't added `haddock` to the tool check toolchain does but
  I'm relying on it being there. This is because it's there normally
  but it's not inconceivable the user might not care about haddock and
  user's toolchain doesn't actually provide it.
* Various versions of Haddock behave in different ways: for example
  Haddock master doesn't output `haddock-util.js` anymore. Unsure how
  to deal with this except letting the user to specify what Haddock
  version they are using (and getting this info from toolchain
  somehow by default somehow).
* There is a lot of repetition of logic when it comes to GHC flags and
  we risk the same situation that we had with
  `haskell_library/haskell_binary`. This sucks but I'm unsure how to
  deal with this nicely: ideally we'd want some sort of "OK let's just
  use the same flags as the package used" with a caveat that not all
  those flags are relevant: we don't care about `-odir` or `-no-link`
  for example. I suspect we could ‘register’ flags we potentially care
  about at build time and propagate them in `HaskellPackageInfo` for
  Haddock's use later (if any). I don't see an elegant way to do this
  except straight up mutating some builder object throughout the
  program though... Nevertheless this needs to be addressed unless we
  want to re-implement a lot of build logic for Haddock.

Closes #34. If we merge this in its current state, I'll create tickets
for some of the above points. May want to address that last one before
even initial merge.

mboes commented on tests/haddock/BUILD:
```diff
@@ -0,0 +1,30 @@
+)
+
+# TODO: Add system library to deps. I can't do it with something like
+# zlib because bazel freaks out that zlib is a non-test dependency
```
Actually, `zlib` *is* only for tests. So should be fine to mark it `testonly`.

mboes commented on tests/binary-with-sysdeps/BUILD:
```diff
@@ -3,6 +3,7 @@ package(default_testonly = 1)
 load("@io_tweag_rules_haskell//haskell:haskell.bzl",
   "haskell_binary",
   "haskell_import",
+
```
extraneous whitespace change.

mboes commented on haskell/haddock.bzl:
```diff
@@ -0,0 +1,165 @@
+
+  static_haddock_outputs = [
+    ctx.actions.declare_file(f, sibling=haddock_interface)
+    for f in ["doc-index.html", "haddock-util.js",
```
Any chance we could avoid this hard coded list of Haddock artifacts?

Fuuzetsu replied:
I don't really see how. We could list these out per-version after #56 but we need these one way or another.

mboes commented:
> ideally we'd want some sort of "OK let's just use the same flags as the package used" with a caveat that not all those flags are relevant

If the non-relevant flags are also non-harmful, then perhaps it's okay to just blindly reuse all build flags as-is?

Fuuzetsu commented:
I opened multiple tickets, addressed the couple of comments and added couple of non-controversial commits which immediately close few of the tickets while I was doing that. Merging.
